### PR TITLE
KEYCLOAK-14412 Keycloak.js should honor configured default scopes when onload 'require-login' is used

### PR DESCRIPTION
**Pull request by thomasdarimont**
We now add the configured default scopes to the login url when onload 'require-login' is used.
Previously, the configured default scopes were ignored when when
`onLoad: 'login-required'` is used in the Keycloak configuration.

### CODE REVIEW

File: adapters/oidc/js/src/main/resources/keycloak.js [hunk 1]
```diff
@@ -438,15 +438,13 @@
...
 
-            var scope;
-            if (options && options.scope) {
-                if (options.scope.indexOf("openid") != -1) {
-                    scope = options.scope;
-                } else {
-                    scope = "openid " + options.scope;
-                }
-            } else {
+            var scope = options && options.scope ? options.scope : kc.scope;
```
Comment (thomasdarimont) [inline adapters/oidc/js/src/main/resources/keycloak.js:441]: If the current options already has a scope configured use that, otherwise use the scope that is configured on the kc (keycloak object) directly, if present.

File: adapters/oidc/js/src/main/resources/keycloak.js [hunk 2]
```diff
@@ -825,6 +823,10 @@
                 configUrl = 'keycloak.json';
             } else if (typeof config === 'string') {
                 configUrl = config;
+            } else if (typeof config === "object") {
+                if (config.scope) {
+                    kc.scope = config.scope;
```
Comment (thomasdarimont) [inline adapters/oidc/js/src/main/resources/keycloak.js:828]: Store default scope from config object into keycloak instance as `kc.scope`

**Review by edewit — DISMISSED** [review]
don't know about what this 'scope' is, but the code looks fine to me

**Review by mhajas — CHANGES_REQUESTED** [review]
Thank you for the changes @thomasdarimont, it looks good to me. However, I am missing some tests for the scope parameter. Looking at the test coverage, we have no tests for the scope parameter so far. Could you please add some? 

You can use something like [this](https://github.com/keycloak/keycloak/blob/master/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java#L388) just change `defaultArguments()` to contain also some scope parameter, or it should be also possible to test `createLoginUrl` directly using [this](https://github.com/keycloak/keycloak/blob/master/testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/util/javascript/JavascriptTestExecutor.java#L251) function.

File: adapters/oidc/js/src/main/resources/keycloak.js [hunk 3]
```diff
@@ -825,6 +823,10 @@
                 configUrl = 'keycloak.json';
             } else if (typeof config === 'string') {
                 configUrl = config;
+            } else if (typeof config === "object") {
```
Comment (jonkoops) [inline adapters/oidc/js/src/main/resources/keycloak.js:826]: Be carefull here as `typeof null === "object"` as well.

Comment (thomasdarimont) [reply adapters/oidc/js/src/main/resources/keycloak.js:826]: Good to know, but isn't this already handled by `if (!config) {` a few lines above?

Comment (jonkoops) [reply adapters/oidc/js/src/main/resources/keycloak.js:826]: Yeah, should be!

File: adapters/oidc/js/src/main/resources/keycloak.js [hunk 4]
```diff
@@ -438,15 +438,13 @@
                 baseUrl = kc.endpoints.authorize();
```
Comment (jonkoops) [inline adapters/oidc/js/src/main/resources/keycloak.js:438]: If we're going to make these changes we also need to make sure that the TypeScript definitions are updated. 

The JSDoc on the `scope` field on the `KeycloakLoginOptions` interface must no longer be private and include some descriptive text for the field itself.

And we'll have to add the same `scope` field to the `KeycloakConfig` interface which is used for the configuration passed into the constructor.

**Review by jonkoops — CHANGES_REQUESTED** [review]

**Comment by jonkoops** [comment]
On a side note, I really think we should consider deprecating the `onLoad` functionality. Although it may appear useful it would be better if the user is forced to call a `checkSSO()` or `login()` method where the options are passed in explicitly.

**Comment by thomasdarimont** [comment]
@jonkoops I revised the PR based on your feedback.
After a looooong evening I eventually got an integration test working....

I noticed that custom `scope` handling is missing from other code-paths as well.
I also added: 
- handling of `scope` in `keycloak.json`. (for which the integration test fails for some reason, it works fine in a local example)
- handling of `scope` in `keycloak.init(...)` the variant `keycloak.login(...)` is already supported.

Any feedback on this is highly appreciated :)

File: adapters/oidc/js/src/main/resources/keycloak.js [hunk 5]
```diff
@@ -432,21 +436,19 @@
...
 
-            var scope;
-            if (options && options.scope) {
-                if (options.scope.indexOf("openid") != -1) {
-                    scope = options.scope;
-                } else {
-                    scope = "openid " + options.scope;
-                }
-            } else {
+            var scope = options && options.scope || kc.scope || config.scope;
```
Comment (thomasdarimont) [inline adapters/oidc/js/src/main/resources/keycloak.js:445]: we check the scope parameter for:
- `options` are the login options
- `kc.scope` potentially derived from `keycloak.json`
- `config.scope` the explicit keycloak configuration.

Comment (thomasdarimont) [reply adapters/oidc/js/src/main/resources/keycloak.js:445]: It would be much better IMHO to create a dedicated config object from `keycloak.json` instead of "parking" the configuration in the `kc` keycloak instance, but this would require a larger refactoring...

File: adapters/oidc/js/src/main/resources/keycloak.d.ts [hunk 6]
```diff
@@ -49,9 +49,22 @@ declare namespace Keycloak {
 		 * Client identifier, example: 'myapp'
 		 */
 		clientId: string;
+
+		/**
+		 * Scope to use for the 'scope' login url parameter.
+		 * The scope 'openid' will be added to the scope if it is missing or undefined.
+		 */
+		scope?: string;
```
Comment (mhajas) [inline adapters/oidc/js/src/main/resources/keycloak.d.ts:57]: Do we need the scope parameter in both `KeycloakConfig` and `KeycloakInitOptions`? `KeycloakConfig` is an equivalent of `AdapterConfig` class for Java adapters and it contains no scope parameter, so I would prefer to remove it, WDYT?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java [hunk 7]
```diff
@@ -400,6 +397,69 @@ public void loginRequiredAction() {
...
+                        .init(initOptions);
+            // This throws exception because when JavascriptExecutor waits for AsyncScript to finish
+            // it is redirected to login page and executor gets no response
+
+            throw new RuntimeException("Probably the login-required OnLoad mode doesn't work, because testExecutor should fail with error that page was redirected.");
+        } catch (WebDriverException ex) {
+            // should happen
+        }
+
+        testExecutor.loginForm(testUser, this::assertOnTestAppUrl)
```
Comment (mhajas) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java:429]: If I am not mistaken, before this line, you can do also something like that.

```java
assertThat(jsDriver.getCurrentUrl(), constainsString("&scope=profile email phone"))
```

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java [hunk 8]
```diff
@@ -400,6 +397,69 @@ public void loginRequiredAction() {
...
+                .init(initOptions, this::assertSuccessfullyLoggedIn)
+        .executeAsyncScript("var callback = arguments[arguments.length - 1];" +
```
Comment (mhajas) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java:431]: This doesn't need to be async script, it is used if we need an output from asynchronous function, in this case you can use simple return like this:
```suggestion
       .executeScript("return window.keycloak.tokenParsed.scope", assertOutputContains("phone"));
```

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java [hunk 9]
```diff
@@ -400,6 +397,69 @@ public void loginRequiredAction() {
...
+        } catch (WebDriverException ex) {
+            // should happen
+        }
+
+        testExecutor.loginForm(testUser, this::assertOnTestAppUrl)
+                .init(initOptions, this::assertSuccessfullyLoggedIn)
+                .executeAsyncScript("var callback = arguments[arguments.length - 1];" +
+                        "callback(window.keycloak.tokenParsed.scope)", assertOutputContains("phone"));
+    }
+
```
Comment (mhajas) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java:462]: Could you add a test also for the `init` function?

**Review by mhajas — CHANGES_REQUESTED** [review]
Thank you very much for the work @thomasdarimont, I really appreciate the effort you made when working on integration tests! 

I added some comments in the code.

Also, I looked into the Jira one more time and it looks like the documentation snippet pasted there is not from Javascript adapter documentation but from Node.js's docs. Anyway, I think this functionality should be added also to JS adapter. However, we need to have all changes we did here also in the documentation before we can merge it.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java (hunk 7 above)
Comment (thomasdarimont) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java:429]: I gave this a try:
```java
assertThat(jsDriver.getCurrentUrl(), containsString("&scope=openid%20profile%20email%20phone"));
```
seems to work.

File: adapters/oidc/js/src/main/resources/keycloak.d.ts (hunk 6 above)
Comment (thomasdarimont) [reply adapters/oidc/js/src/main/resources/keycloak.d.ts:57]: I didn't think of that - yes, you are right. I'll remove the prop again from Keycloak config.

**Comment by thomasdarimont** [comment]
@mhajas thanks for your suggestions! I think it would be best to streamline the ways scopes can be configured in the JS adapter. I still think a good place for this configuration is `initOptions` that can be used like this:
```
  keycloak.init({
                onLoad: 'login-required', 
                checkLoginIframe: true, 
                checkLoginIframeInterval: 1, 
                pkceMethod: 'S256', 
                scope: 'profile email phone'
              })
```
I can also think of support for `loginOptions`:
```
keycloak.login({
   ...,
  scope:"profile email phone"
}) 
```
but now I'm not sure anymore, whether it is a good idea to have 2 ways for configuring scopes.
Do you think users might want to have a generic configuration in `keycloak.init(initOptions)` and override this with the `keycloak.login({loginOptions})`?

Also I'm still fighting with the tests. It is really hard to find out what's wrong if I cannot debug the generated JS code in a browser. Is it somehow possible to attach a chrome debugger to headless webdriver browser instance? Perhaps in combination with the `debugger` keyword?

**Comment by thomasdarimont** [comment]
I think I eventually got something working...

The scope of the issue (phun intended) should be changed to:
`KEYCLOAK-14412 Keycloak.js should honor scopes configured in initOptions and loginOptions`

**Comment by mhajas** [comment]
> but now I'm not sure anymore, whether it is a good idea to have 2 ways for configuring scopes.
Do you think users might want to have a generic configuration in keycloak.init(initOptions) and override this with the keycloak.login({loginOptions})?

I would say it does, as when we are using `onLoad: login-required` option, we are not able to add scopes to the login request at the moment.  

> Also I'm still fighting with the tests. It is really hard to find out what's wrong if I cannot debug the generated JS code in a browser. Is it somehow possible to attach a chrome debugger to headless webdriver browser instance? 

As I understand it the code we are putting into `executeScript` or `executeAsyncScript` is basically executed in the console of the browser, so I am not sure if it is possible to debug it. However, you can run the test in a non-headless browser using following props: `-Djs.browser=chrome -Djs.chromeArguments=""`.

> Perhaps in combination with the debugger keyword?
I have never heard of the `debugger` keyword, feel free to try it. It would be really helpful if it works.

File: adapters/oidc/js/src/main/resources/keycloak.js [hunk 10]
```diff
@@ -276,6 +280,7 @@
                 };
 
                 var options = {};
+                options.scope = kc.scope;
```
Comment (mhajas) [inline adapters/oidc/js/src/main/resources/keycloak.js:283]: I think this is not necessary. `kc.skope` will be used within the `login` function directly.

Comment (thomasdarimont) [reply adapters/oidc/js/src/main/resources/keycloak.js]: Yep, this was a left over from my experiments.

**Comment by thomasdarimont** [comment]
@mhajas thanks for your tips!

File: adapters/oidc/js/src/main/resources/keycloak.js (hunk 10 above)
Comment (thomasdarimont) [reply adapters/oidc/js/src/main/resources/keycloak.js]: Just removed this line. Build is running.

**Comment by thomasdarimont** [comment]
I rebased the two commits into one.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java (hunk 9 above)
Comment (thomasdarimont) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java:452]: There are now two tests, one for `initOptions` and another one `loginOptions`.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java [hunk 11]
```diff
@@ -400,6 +398,58 @@ public void loginRequiredAction() {
...
+            testExecutor.configure()
+                        .init(defaultArguments());
+        } catch (WebDriverException ex) {
+            // should happen
+        }
+
+        JSObjectBuilder loginOptions = JSObjectBuilder.create().add("scope", "profile email phone");
+
+        testExecutor.login(loginOptions.build(), (JavascriptStateValidator) (driver, output, events) -> {
+            assertThat(jsDriver.getCurrentUrl(), containsString("&scope=openid%20profile%20email%20phone"));
```
Comment (mhajas) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java:449]: The JavascriptAdapter tests are a little bit confusing in this. They are mixing two variable names `driver` and `jsDriver`. Out testsuite has two browsers because JS adapter doesn't work in Htmlunit, which is the default browser for majority of tests (it is default because it is fast). So we added `jsDriver` as the second browser, which defaults to Phantomjs. This browser is able to work with JS adapter. 

Now the `JavascriptStateValidator` is there for checking the state after some Javascript piece of code was executed. It has three parameters `(driver, output, events)`. At the moment, the `driver` parameter is maybe a little bit redundant here as we always use `jsDriver`, but this can change in the future. So it would be better to use it here like this:
```suggestion
            assertThat(driver.getCurrentUrl(), containsString("&scope=openid%20profile%20email%20phone"));
```

this way you use a webdriver provided by testExecutor and function (in this case `login()`) can decide what browser to use for validating the condition.


One more thing, you don't need to use the `testExecutor` for everything, the `jsDriver` (do not use `driver` outside of) is available even outside of testExecutor and it is perfectly ok to use it.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java [hunk 12]
```diff
@@ -400,6 +398,58 @@ public void loginRequiredAction() {
...
+        try {
+            testExecutor.init(initOptions);
+            // This throws exception because when JavascriptExecutor waits for AsyncScript to finish
+            // it is redirected to login page and executor gets no response
+
+            throw new RuntimeException("Probably the login-required OnLoad mode doesn't work, because testExecutor should fail with error that page was redirected.");
+        } catch (WebDriverException ex) {
+            // should happen
+        }
+
```
Comment (mhajas) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java:424]: You can put this here as well:
```suggestion
assertThat(jsDriver.getCurrentUrl(), containsString("&scope=openid%20profile%20email%20phone"));
```

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java (hunk 11 above)
Comment (mhajas) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java]: Hopefully, it is understandable and I don't just confuse you more. I like our JS tests, but lots of people are struggling with them. It is almost done though,  just some nitpicks to fix

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java (hunk 12 above)
Comment (thomasdarimont) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java:424]: This doesn't work here:
```
Expected :a string containing "&scope=openid%20profile%20email%20phone"
Actual   :"https://localhost:8543/auth/realms/test/protocol/openid-connect/auth?client_id=account&redirect_uri=https%3A%2F%2Flocalhost%3A8543%2Fauth%2Frealms%2Ftest%2Faccount%2Flogin-redirect&state=0%2F4c2c6895-ca59-491c-b874-2d440ee84732&response_type=code&scope=openid"
```

The following test for the `phone` claim works though.

**Comment by thomasdarimont** [comment]
The github builds fail with `Unexpected I/O error` - the tests work fine locally.

File: adapters/oidc/js/src/main/resources/keycloak.js (hunk 5 above)
Comment (jonkoops) [reply adapters/oidc/js/src/main/resources/keycloak.js]: Yeah, this is certainly something that is hard to follow as it is now. Should be re-factored in the future.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java (hunk 12 above)
Comment (mhajas) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java:424]: Weird. This works for me, are you sure you used the exact line I pasted here?

**Review by mhajas — CHANGES_REQUESTED** [review]

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java [hunk 13]
```diff
@@ -400,6 +398,58 @@ public void loginRequiredAction() {
...
+    /**
+     * Test for scope handling via {@code loginOptions}: <pre>{@code
+     * Keycloak keycloak = new Keycloak(); keycloak.login({.... scope: "profile email phone"})
+     * }</pre>
+     * See KEYCLOAK-14412
+     */
+    @Test
+    public void testScopeInLoginOptionsShouldBeConsideredByLoginUrl() {
+
+        try {
```
[the remark this answers is not in the document]
Comment (thomasdarimont) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java:439]: Okay, I'll remove the block.

File: testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/util/javascript/JavascriptTestExecutor.java [hunk 14]
```diff
@@ -39,13 +39,17 @@ protected JavascriptTestExecutor(WebDriver driver, OIDCLogin loginPage) {
     }
 
     public JavascriptTestExecutor login() {
-        return login(null, null);
+        return login((JSObjectBuilder)null, null);
     }
     
     public JavascriptTestExecutor login(JavascriptStateValidator validator) {
-        return login(null, validator);
+        return login((JSObjectBuilder)null, validator);
```
[the remark this answers is not in the document]
Comment (thomasdarimont) [reply testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/util/javascript/JavascriptTestExecutor.java:46]: k

File: testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/util/javascript/JavascriptTestExecutor.java [hunk 15]
```diff
@@ -39,13 +39,17 @@ protected JavascriptTestExecutor(WebDriver driver, OIDCLogin loginPage) {
     }
 
     public JavascriptTestExecutor login() {
-        return login(null, null);
+        return login((JSObjectBuilder)null, null);
```
[the remark this answers is not in the document]
Comment (thomasdarimont) [reply testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/util/javascript/JavascriptTestExecutor.java:42]: k

**Comment by thomasdarimont** [comment]
Yes this is the test method... 
The test fails if I uncomment: 
```
//     assertThat(driver.getCurrentUrl()....
```

```
    @Test
    public void testScopeInInitOptionsShouldBeConsideredByLoginUrl() {

        JSObjectBuilder initOptions = defaultArguments()
                .loginRequiredOnLoad()
                // phone is optional client scope
                .add("scope", "profile email phone");

        try {
            testExecutor.init(initOptions);
            // This throws exception because when JavascriptExecutor waits for AsyncScript to finish
            // it is redirected to login page and executor gets no response

            throw new RuntimeException("Probably the login-required OnLoad mode doesn't work, because testExecutor should fail with error that page was redirected.");
        } catch (WebDriverException ex) {
            // should happen
        }

   //     assertThat(driver.getCurrentUrl(), containsString("&scope=openid%20profile%20email%20phone"));
        testExecutor.loginForm(testUser, this::assertOnTestAppUrl)
                    .init(initOptions, this::assertSuccessfullyLoggedIn)
                    .executeScript("return window.keycloak.tokenParsed.scope", assertOutputContains("phone"));
    }
```

**Comment by mhajas** [comment]
> Yes this is the test method...
> The test fails if I uncomment:
> 
> ```
> //     assertThat(driver.getCurrentUrl()....
> ```
> 
> ```
>     @Test
>     public void testScopeInInitOptionsShouldBeConsideredByLoginUrl() {
> 
>         JSObjectBuilder initOptions = defaultArguments()
>                 .loginRequiredOnLoad()
>                 // phone is optional client scope
>                 .add("scope", "profile email phone");
> 
>         try {
>             testExecutor.init(initOptions);
>             // This throws exception because when JavascriptExecutor waits for AsyncScript to finish
>             // it is redirected to login page and executor gets no response
> 
>             throw new RuntimeException("Probably the login-required OnLoad mode doesn't work, because testExecutor should fail with error that page was redirected.");
>         } catch (WebDriverException ex) {
>             // should happen
>         }
> 
>    //     assertThat(driver.getCurrentUrl(), containsString("&scope=openid%20profile%20email%20phone"));
>         testExecutor.loginForm(testUser, this::assertOnTestAppUrl)
>                     .init(initOptions, this::assertSuccessfullyLoggedIn)
>                     .executeScript("return window.keycloak.tokenParsed.scope", assertOutputContains("phone"));
>     }
> ```

Sorry, I probably haven't explained it properly before. In this case, there should be `jsDriver` not `driver`. `driver` must be used only in case it is within JavascriptStateValidator lambda function.

**Comment by mhajas** [comment]
@thomasdarimont I think the line we were discussing can be added to the tests if you use `jsDriver`. Anyway, the functionality should be tested even in this case, so I am okay with the test as it is. Could you rebase the PR? Then it should be ready to merge.

**Comment by thomasdarimont** [comment]
@mhajas Thanks for your feedback. I just rebased the PR.

**Review by mhajas — DISMISSED** [review]

**Comment by mhajas** [comment]
@thomasdarimont Seems like Travis found some compilation error. Could you have a look?

**Comment by mhajas** [comment]
Hey @thomasdarimont, do you plan to continue with this PR?

File: adapters/oidc/js/src/main/resources/keycloak.js (hunk 4 above)
Comment (thomasdarimont) [reply adapters/oidc/js/src/main/resources/keycloak.js:437]: I eventually found some time to get back to this...
@jonkoops can you point me to the TypeScript definitions that need to be changed for this?

**Comment by thomasdarimont** [comment]
@mhajas I eventually found the time to give this another spin.

I fixed the compiler error in the `org.keycloak.testsuite.javascript.JavascriptAdapterTest#testScopeInInitOptionsShouldBeConsideredByLoginUrl` method.
Unfortunately I cannot get the test method `testScopeInInitOptionsShouldBeConsideredByLoginUrl` to pass, since the generated login url for the account client does not use the scope `profile email phone` given in via initOptions, it just uses `openid`.

However if I test the `keycloak.js` from this PR manually with the following init options in an example app it works.
```
keycloak.init({onLoad: 'login-required', checkLoginIframe: true, checkLoginIframeInterval: 1, pkceMethod: 'S256', scope: "openid profile phone"})
```

Do you have any idea what might be wrong here? I suspect that the test-setup does something special for the account client when `testExecutor.loginForm(...)` is used.

File: adapters/oidc/js/src/main/resources/keycloak.js (hunk 4 above)
Comment (jonkoops) [reply adapters/oidc/js/src/main/resources/keycloak.js:437]: Sure, you can find them in the same directory right [here](https://github.com/keycloak/keycloak/blob/master/adapters/oidc/js/src/main/resources/keycloak.d.ts).

Comment (jonkoops) [reply adapters/oidc/js/src/main/resources/keycloak.js:437]: Hmm, I think this change perhaps has been removed? I don't remember the original code so I guess we can resolve this.

**Review by mhajas — DISMISSED** [review]
@thomasdarimont The problem with the failing test is probably on you local machine. It might be caused by wrong keycloak.js file picked by the tests. I think these steps should solve your problem: 
1. Rebuild `distribution/adapters/js-adapter-zip` or simply run `mvn clean install -Pdistribution -DskipTests` from keycloak-parent
2. Rebuild `testsuite/integration-arquillian/servers/auth-server/services/testsuite-providers
3. Check if keycloak.js in testsuite-providers/target/classes/javascript/keycloak.js and in resuilting jar file contains your changes.
3. Run the tests, now it should pass 

Let me know if you are still not able to make it pass. To me this seems ok now, I think it should be ready to merge.

**Comment by thomasdarimont** [comment]
@mhajas thanks for the tipps! Just gave it a spin and it works now :)

**Review by mhajas — DISMISSED** [review]
Thanks for confirmation @thomasdarimont. @abstractj I believe this should be ready for merge.

**Review by abstractj — APPROVED** [review]

**Review by abstractj — CHANGES_REQUESTED** [review]
@thomasdarimont could you please rebase? Once it's done we're ready to merge.

**Comment by thomasdarimont** [comment]
@abstractj thanks for looking into this, I just rebased the PR on current master.

**Review by mhajas — APPROVED** [review]

**Comment by tiagosimao** [comment]
Hi guys, just an awkward heads-up.

The Keycloak JS lib has a similar problem with the idpHint option that exists in login(). There seems to be no way to have the idpHint Feature when using the "login-required" flag.

I just commented this here because your issue tracker now requires a Red Hat account, which is something I'm not willing to have/use. I was otherwise willing to contribute.

**Review by mhajas — APPROVED** [review]
@abstractj the changes look good to me. I tested locally, I also tried to rebase to the current master and everything is working as expected. From my POV this is ready for merge.

**Review by abstractj — APPROVED** [review]
Thanks @mhajas reviewing the changes, they LGTM

@thomasdarimont thank you and my apologies for taking so long.
